include/libgit4cpp/Error.h

```diff
@@ -25,36 +25,114 @@
 #ifndef LIBGIT4CPP_ERROR_H_
 #define LIBGIT4CPP_ERROR_H_
 
-#include <stdexcept>
+#include <system_error>
 
 namespace git {
 
+class git_category_impl : public std::error_category {
+    public:
+        virtual const char* name() const noexcept;
+        virtual std::string message(int ev) const;
+};
+
+const std::error_category& git_category();
+
 /**
- * An exception class carrying an error message.
+ * An exception class carrying an error code and message.
  *
  * Error is used as the standard exception class throughout libgit4cpp.
  * It can be used directly or inherited from.
  *
  * \code
  * try
  * {
- *     throw git::Error("An error has occurred");
+ *     throw git::Error(errorcode, "An error has occurred");
+ *     // or throw git::Error("Another error occurred");
  * }
  * catch (const git::Error& e)
  * {
  *     std::cerr << e.what() << "\n";
  * }
  * \endcode
  *
+ * If no error code is specified GIT_EUSER is used.
+ *
  * \note
- * git::Error is derived from std::runtime_error. It can therefore be caught by
- * `catch (const std::exception&)`, `catch (const std::runtime_error&)`, and
- * `catch (const git::Error&)`.
+ * git::Error is derived from std::system_error. It can therefore be caught by
+ * `catch (const std::exception&)`, `catch (const std::runtime_error&)`,
+ * `catch (const std::system_error&)`, and `catch (const git::Error&)`.
  */
-class Error : public std::runtime_error
+class Error : public std::system_error
 {
 public:
-    using std::runtime_error::runtime_error;
+    using std::system_error::system_error;
+    Error(int ev)
+        : std::system_error(ev, git_category())
+        { }
+    Error(int ev, const std::string& what)
```

Comment:
The order of parameters matches `std::system_error`, so this is probably the best way to go. But just to mention it: In `doocs::`, `hlc::` and probably other contexts we have `ErrorCode` classes that take the parameters in the inverse order, and those `ErrorCode` classes are derived from `Error` classes that just contain a string. But I would leave it as implemented here.

Reply:
> In doocs::, hlc:: and probably other contexts we have ErrorCode classes that take the parameters in the inverse order

:sob: Why would one write `ErrorCode` that does not inherit from `std::system_error`.
Probably at least in the longer run, maybe that other libs should be fixed to use a "standard conforming" error code object.